File: backend/agent/tools/browser.py
from __future__ import annotations

import asyncio
import datetime
import urllib.parse
from typing import Any, Dict

from agent.tools.base import Tool
from agent.tools.input import _get_screenshot_dir, _cleanup_old_screenshots


# ── persistent interactive session (shared by browser_click / browser_fill) ──
# Keeps one headless page alive across calls so clicks and fills compound,
# unlike the stateless read/screenshot tools that open a fresh browser each time.
_SESSION: Dict[str, Any] = {"pw": None, "browser": None, "page": None}


async def _close_session() -> None:
	"""Close the persistent browser and playwright instances."""
	browser = _SESSION.get("browser")
	if browser is not None:
		try:
			await browser.close()
		except Exception:
			pass
	pw = _SESSION.get("pw")
	if pw is not None:
		try:
			await pw.stop()
		except Exception:
			pass
	_SESSION.update(pw=None, browser=None, page=None)


async def _get_page():
	"""Return the live interactive page, creating the browser lazily."""
	from playwright.async_api import async_playwright

	page = _SESSION.get("page")
	if page is not None and not page.is_closed():
		return page
	# Clean up a stale session before creating a new one.
	await _close_session()
	pw = await async_playwright().start()
	browser = await pw.chromium.launch(headless=True)
	ctx = await browser.new_context(locale="en-US")
	page = await ctx.new_page()
	_SESSION.update(pw=pw, browser=browser, page=page)
	return page


def _selector(raw: str) -> str:
	"""Accept a CSS/XPath selector or plain text.

	If the string contains no selector metacharacters it's treated as visible
	text and matched via Playwright's text engine. Pass an explicit
	`text=...` or a CSS selector to be unambiguous.
	"""
	raw = raw.strip()
	if raw.startswith(("#", ".", "[", "//")) or raw.startswith("text="):
		return raw
	if any(c in raw for c in "#.[]>=:") or "//" in raw:
		return raw
	# looks like human text → match by visible text
	return f"text={raw}"


async def _page_snapshot(page) -> str:
	try:
		text = await page.evaluate("() => document.body ? document.body.innerText.slice(0, 800) : ''")
	except Exception:
		text = ""
	url = page.url
	out = f"url: {url}"
	if text and text.strip():
		out += "\n" + text.strip()
	return out[:1500]


class BrowserOpenTool(Tool):
	name = "browser_open"
	description = (
		"Open a URL in the user's browser. "
		"Does not read or control the page. For reading and checking web pages use browser_read or browser_screenshot."
	)
	params = {"url": {"type": "string", "description": "URL"}}

	async def execute(self, args: Dict[str, Any]) -> str:
		url = args.get("url", "")
		try:
			import webbrowser
			await asyncio.to_thread(webbrowser.open, url)
			return f"opened: {url}"
		except Exception as e:
			return f"error: {e}"


class BrowserReadTool(Tool):
	name = "browser_read"
	description = (
		"Read page content via Playwright: text and list of links. "
		"Use for site navigation, page checks and data extraction without opening a window for the user."
	)
	params = {"url": {"type": "string", "description": "URL"}}

	async def execute(self, args: Dict[str, Any]) -> str:
		url = args.get("url", "")
		try:
			from playwright.async_api import async_playwright
			async with async_playwright() as pw:
				browser = await pw.chromium.launch(headless=True)
				ctx = await browser.new_context(locale="en-US")
				page = await ctx.new_page()
				await page.goto(url, timeout=20000)
				for sel in [
					'button:has-text("Accept all")',
					'button:has-text("Reject all")',
					'button[aria-label*="Accept"]',
					'#L2AGLb',
					'button:has-text("Agree")',
				]:
					try:
						await page.click(sel, timeout=1500)
						break
					except Exception:
						pass
				try:
					await page.wait_for_load_state("networkidle", timeout=5000)
				except Exception:
					pass
				data = await page.evaluate("""
					() => {
						const text = document.body.innerText.slice(0, 2000);
						const links = [...document.querySelectorAll('a[href]')]
							.map(a => ({text: (a.innerText || a.title || '').trim().slice(0, 80), url: a.href}))
							.filter(l => l.text && l.url && !l.url.startsWith('javascript') && !l.url.startsWith('mailto'))
							.slice(0, 40);
						return {text, links};
					}
				""")
				await browser.close()
			out = data["text"]
			if data["links"]:
				out += "\n\nLinks:\n" + "\n".join(f"• {l['text']}: {l['url']}" for l in data["links"])
			return out[:3000]
		except ImportError:
			return "error: pip install playwright && playwright install chromium"
		except Exception as e:
			return f"error: {e}"


class YouTubeSearchTool(Tool):
	name = "youtube_search"
	description = (
		"Search for videos on YouTube. Returns a list of videos with direct links. "
		"Use instead of google_search for video search."
	)
	params = {"query": {"type": "string", "description": "Search query"}}

	async def execute(self, args: Dict[str, Any]) -> str:
		query = args.get("query", "")
		try:
			from playwright.async_api import async_playwright
			async with async_playwright() as pw:
				browser = await pw.chromium.launch(headless=True)
				try:
					ctx = await browser.new_context(locale="en-US")
					page = await ctx.new_page()
					await page.goto(
						f"https://www.youtube.com/results?search_query={urllib.parse.quote(query)}",
						timeout=20000,
					)
					for sel in [
						'button:has-text("Accept all")',
						'button:has-text("Reject all")',
						'button[aria-label*="Accept"]',
					]:
						try:
							await page.click(sel, timeout=2000)
							break
						except Exception:
							pass
					try:
						await page.wait_for_selector("ytd-video-renderer", timeout=8000)
					except Exception:
						pass
					results = await page.evaluate("""
						() => {
							const items = document.querySelectorAll('ytd-video-renderer');
							return [...items].slice(0, 8).map(item => {
								const a = item.querySelector('a#video-title');
								const meta = item.querySelector('#metadata-line');
								return {
									title: (a?.textContent || '').trim(),
									url: a?.href || '',
									meta: (meta?.textContent || '').trim().replace(/\\s+/g, ' ')
								};
							}).filter(r => r.title && r.url);
						}
					""")
				finally:
					await browser.close()
			if not results:
				return "no results"
			return "\n".join(
				f"{i+1}. {r['title']}{(' · ' + r['meta']) if r['meta'] else ''}\n   {r['url']}"
				for i, r in enumerate(results)
			)
		except ImportError:
			return "error: pip install playwright && playwright install chromium"
		except Exception as e:
			return f"error: {e}"


class BrowserScreenshotTool(Tool):
	name = "browser_screenshot"
	description = "Take a screenshot of a web page in the background via Playwright and return the file path."
	params = {"url": {"type": "string", "description": "URL"}}

	async def execute(self, args: Dict[str, Any]) -> str:
		url = args.get("url", "")
		try:
			from playwright.async_api import async_playwright
			screenshot_dir = _get_screenshot_dir()
			_cleanup_old_screenshots(screenshot_dir, max_age_seconds=300)
			name = screenshot_dir / f"browser_{datetime.datetime.now():%Y%m%d_%H%M%S}.png"
			async with async_playwright() as pw:
				browser = await pw.chromium.launch(headless=True)
				page = await browser.new_page()
				await page.goto(url, timeout=20000)
				await page.screenshot(path=str(name), full_page=True)
				await browser.close()
			return f"saved: {name}"
		except ImportError:
			return "error: pip install playwright && playwright install chromium"
		except Exception as e:
			return f"error: {e}"


class BrowserClickTool(Tool):
	name = "browser_click"
	description = (
		"Click an element on the interactive browser page (Playwright). "
		"Optionally navigate to a URL first. "
		"selector: a CSS/XPath selector or plain visible text. "
		"The session persists, so clicks and fills build on each other."
	)
	params = {
		"selector": {"type": "string", "description": "CSS/XPath selector or visible text to click"},
		"url": {"type": "string", "description": "URL to navigate to first (optional)"},
		"timeout": {"type": "integer", "description": "Timeout in seconds for the action (default 15)"},
	}

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = ["selector"]
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		selector = str(args.get("selector", "")).strip()
		if not selector:
			return "error: selector is required"
		url = str(args.get("url", "")).strip()
		timeout_ms = min(int(args.get("timeout", 15)), 60) * 1000
		try:
			page = await _get_page()
			if url:
				await page.goto(url, timeout=timeout_ms)
			await page.click(_selector(selector), timeout=timeout_ms)
			try:
				await page.wait_for_load_state("networkidle", timeout=5000)
			except Exception:
				pass
			snap = await _page_snapshot(page)
			return f"clicked: {selector}\n{snap}"
		except ImportError:
			return "error: pip install playwright && playwright install chromium"
		except Exception as e:
			return f"error: {e}"


class BrowserFillTool(Tool):
	name = "browser_fill"
	description = (
		"Fill an input field on the interactive browser page (Playwright). "
		"Optionally navigate to a URL first and/or press Enter after filling. "
		"selector: a CSS/XPath selector or plain visible text/label. "
		"The session persists across calls."
	)
	params = {
		"selector": {"type": "string", "description": "CSS/XPath selector or label/placeholder text of the field"},
		"value": {"type": "string", "description": "Text to type into the field"},
		"url": {"type": "string", "description": "URL to navigate to first (optional)"},
		"submit": {"type": "boolean", "description": "Press Enter after filling (optional)"},
		"timeout": {"type": "integer", "description": "Timeout in seconds for the action (default 15)"},
	}

	def tool_definition(self) -> dict:
		d = super().tool_definition()
		d["function"]["parameters"]["required"] = ["selector", "value"]
		return d

	async def execute(self, args: Dict[str, Any]) -> str:
		selector = str(args.get("selector", "")).strip()
		if not selector:
			return "error: selector is required"
		value = str(args.get("value", ""))
		url = str(args.get("url", "")).strip()
		submit = bool(args.get("submit", False))
		timeout_ms = min(int(args.get("timeout", 15)), 60) * 1000
		try:
			page = await _get_page()
			if url:
				await page.goto(url, timeout=timeout_ms)
			await page.fill(_selector(selector), value, timeout=timeout_ms)
			if submit:
				await page.press(_selector(selector), "Enter")
				try:
					await page.wait_for_load_state("networkidle", timeout=5000)
				except Exception:
					pass
			snap = await _page_snapshot(page)
			return f"filled: {selector}\n{snap}"
		except ImportError:
			return "error: pip install playwright && playwright install chromium"
		except Exception as e:
			return f"error: {e}"

File: agent/tools/browser.go
package tools

import (
	"fmt"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/mxschmitt/playwright-go"
)

var (
	pw      *playwright.Playwright
	browser playwright.Browser
	page    playwright.Page
	sessMu  sync.Mutex
)

func closeSession() {
	sessMu.Lock()
	defer sessMu.Unlock()
	if page != nil {
		page.Close()
		page = nil
	}
	if browser != nil {
		browser.Close()
		browser = nil
	}
	if pw != nil {
		pw.Stop()
		pw = nil
	}
}

func getPage() (playwright.Page, error) {
	sessMu.Lock()
	defer sessMu.Unlock()

	if page != nil && !page.IsClosed() {
		return page, nil
	}

	closeSessionLocked()

	var err error
	pw, err = playwright.Run()
	if err != nil {
		return nil, fmt.Errorf("playwright: %w", err)
	}

	browser, err = pw.Chromium.Launch(playwright.BrowserTypeLaunchOptions{Headless: playwright.Bool(true)})
	if err != nil {
		pw.Stop()
		pw = nil
		return nil, fmt.Errorf("playwright launch: %w", err)
	}

	ctx, err := browser.NewContext(playwright.BrowserNewContextOptions{Locale: playwright.String("en-US")})
	if err != nil {
		browser.Close()
		browser = nil
		pw.Stop()
		pw = nil
		return nil, fmt.Errorf("playwright context: %w", err)
	}

	page, err = ctx.NewPage()
	if err != nil {
		browser.Close()
		browser = nil
		pw.Stop()
		pw = nil
		return nil, fmt.Errorf("playwright page: %w", err)
	}

	return page, nil
}

func closeSessionLocked() {
	if page != nil {
		page.Close()
		page = nil
	}
	if browser != nil {
		browser.Close()
		browser = nil
	}
	if pw != nil {
		pw.Stop()
		pw = nil
	}
}

func pwSelector(raw string) string {
	raw = strings.TrimSpace(raw)
	if strings.HasPrefix(raw, "#") || strings.HasPrefix(raw, ".") ||
		strings.HasPrefix(raw, "[") || strings.HasPrefix(raw, "//") ||
		strings.HasPrefix(raw, "text=") {
		return raw
	}
	for _, c := range "#.[]>=:" {
		if strings.ContainsRune(raw, c) {
			return raw
		}
	}
	if strings.Contains(raw, "//") {
		return raw
	}
	return "text=" + raw
}

func pageSnapshot(p playwright.Page) string {
	text := ""
	result, err := p.Evaluate("() => document.body ? document.body.innerText.slice(0, 800) : ''")
	if err == nil {
		if s, ok := result.(string); ok {
			text = s
		}
	}

	pageURL := p.URL()
	out := "url: " + pageURL
	if strings.TrimSpace(text) != "" {
		out += "\n" + strings.TrimSpace(text)
	}
	if len(out) > 1500 {
		out = out[:1500]
	}
	return out
}

var cookieConsentSelectors = []string{
	`button:has-text("Accept all")`,
	`button:has-text("Reject all")`,
	`button[aria-label*="Accept"]`,
	`#L2AGLb`,
	`button:has-text("Agree")`,
}

func dismissCookieConsent(p playwright.Page) {
	for _, sel := range cookieConsentSelectors {
		err := p.Click(sel, playwright.PageClickOptions{Timeout: playwright.Float(1500)})
		if err == nil {
			return
		}
	}
}

func newBrowserPage() (playwright.Browser, playwright.Page, func(), error) {
	var err error
	tempPW, err := playwright.Run()
	if err != nil {
		return nil, nil, nil, fmt.Errorf("playwright init: %w", err)
	}

	tempBrowser, err := tempPW.Chromium.Launch(playwright.BrowserTypeLaunchOptions{Headless: playwright.Bool(true)})
	if err != nil {
		tempPW.Stop()
		return nil, nil, nil, fmt.Errorf("playwright launch: %w", err)
	}

	ctx, err := tempBrowser.NewContext(playwright.BrowserNewContextOptions{Locale: playwright.String("en-US")})
	if err != nil {
		tempBrowser.Close()
		tempPW.Stop()
		return nil, nil, nil, fmt.Errorf("playwright context: %w", err)
	}

	tempPage, err := ctx.NewPage()
	if err != nil {
		tempBrowser.Close()
		tempPW.Stop()
		return nil, nil, nil, fmt.Errorf("playwright page: %w", err)
	}

	cleanup := func() {
		tempBrowser.Close()
		tempPW.Stop()
	}

	return tempBrowser, tempPage, cleanup, nil
}

// ── BrowserOpenTool ───────────────────────────────────────────────────────────

type BrowserOpenTool struct{}

func (t *BrowserOpenTool) Name() string { return "browser_open" }

func (t *BrowserOpenTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Open a URL in the system browser.",
		Params: map[string]any{
			"url": prop("string", "URL to open"),
		},
		Required: []string{"url"},
	}
}

func (t *BrowserOpenTool) Execute(args map[string]any) Result {
	rawURL := strings.TrimSpace(getStr(args, "url"))
	if !isSSRFSafeURL(rawURL) {
		return R("error: URL is blocked (SSRF or file scheme)")
	}

	u, err := url.Parse(rawURL)
	if err != nil || (u.Scheme != "http" && u.Scheme != "https") {
		return R("error: invalid URL")
	}

	cmd := exec.Command("rundll32", "url.dll,FileProtocolHandler", rawURL)
	if err := cmd.Start(); err != nil {
		return R("error: " + err.Error())
	}

	return R("opened: " + rawURL)
}

// ── BrowserReadTool ───────────────────────────────────────────────────────────

type BrowserReadTool struct{}

func (t *BrowserReadTool) Name() string { return "browser_read" }

func (t *BrowserReadTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Open a URL in a headless browser and read its text and links.",
		Params: map[string]any{
			"url": prop("string", "URL to read"),
		},
		Required: []string{"url"},
	}
}

func (t *BrowserReadTool) Execute(args map[string]any) Result {
	rawURL := strings.TrimSpace(getStr(args, "url"))
	if !isSSRFSafeURL(rawURL) {
		return R("error: URL is blocked (SSRF or file scheme)")
	}

	_, tempPage, cleanup, err := newBrowserPage()
	if err != nil {
		return R("error: playwright not available — run `playwright install chromium`")
	}
	defer cleanup()

	if _, err := tempPage.Goto(rawURL, playwright.PageGotoOptions{Timeout: playwright.Float(20000)}); err != nil {
		return R("error: " + err.Error())
	}

	dismissCookieConsent(tempPage)

	tempPage.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
		State:   playwright.LoadStateNetworkidle,
		Timeout: playwright.Float(5000),
	})

	result, err := tempPage.Evaluate(`() => {
		const text = document.body.innerText.slice(0, 2000);
		const links = [...document.querySelectorAll('a[href]')]
			.map(a => ({text: (a.innerText || a.title || '').trim().slice(0, 80), url: a.href}))
			.filter(l => l.text && l.url && !l.url.startsWith('javascript') && !l.url.startsWith('mailto'))
			.slice(0, 40);
		return {text, links};
	}`)
	if err != nil {
		return R("error: " + err.Error())
	}

	data, ok := result.(map[string]interface{})
	if !ok {
		return R("error: unexpected page content")
	}

	var out string
	if t, ok := data["text"].(string); ok {
		out = t
	}

	if links, ok := data["links"].([]interface{}); ok && len(links) > 0 {
		out += "\n\nLinks:\n"
		for _, l := range links {
			if lm, ok := l.(map[string]interface{}); ok {
				txt := ""
				linkURL := ""
				if v, ok := lm["text"].(string); ok {
					txt = v
				}
				if v, ok := lm["url"].(string); ok {
					linkURL = v
				}
				if txt != "" && linkURL != "" {
					out += fmt.Sprintf("- %s: %s\n", txt, linkURL)
				}
			}
		}
	}

	if len(out) > 3000 {
		out = out[:3000]
	}
	return R(out)
}

// ── YouTubeSearchTool ─────────────────────────────────────────────────────────

type YouTubeSearchTool struct{}

func (t *YouTubeSearchTool) Name() string { return "youtube_search" }

func (t *YouTubeSearchTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Search YouTube and return video results.",
		Params: map[string]any{
			"query": prop("string", "Search query"),
		},
		Required: []string{"query"},
	}
}

func (t *YouTubeSearchTool) Execute(args map[string]any) Result {
	query := strings.TrimSpace(getStr(args, "query"))
	if query == "" {
		return R("error: query is required")
	}

	_, tempPage, cleanup, err := newBrowserPage()
	if err != nil {
		return R("error: playwright not available — run `playwright install chromium`")
	}
	defer cleanup()

	searchURL := "https://www.youtube.com/results?search_query=" + url.QueryEscape(query)
	if _, err := tempPage.Goto(searchURL, playwright.PageGotoOptions{Timeout: playwright.Float(20000)}); err != nil {
		return R("error: " + err.Error())
	}

	for _, sel := range cookieConsentSelectors {
		err := tempPage.Click(sel, playwright.PageClickOptions{Timeout: playwright.Float(2000)})
		if err == nil {
			break
		}
	}

	tempPage.WaitForSelector("ytd-video-renderer", playwright.PageWaitForSelectorOptions{Timeout: playwright.Float(8000)})

	result, err := tempPage.Evaluate(`() => {
		const items = document.querySelectorAll('ytd-video-renderer');
		return [...items].slice(0, 8).map(item => {
			const a = item.querySelector('a#video-title');
			const meta = item.querySelector('#metadata-line');
			return {
				title: (a?.textContent || '').trim(),
				url: a?.href || '',
				meta: (meta?.textContent || '').trim().replace(/\s+/g, ' ')
			};
		}).filter(r => r.title && r.url);
	}`)
	if err != nil {
		return R("error: " + err.Error())
	}

	items, ok := result.([]interface{})
	if !ok || len(items) == 0 {
		return R("no results")
	}

	var b strings.Builder
	for i, item := range items {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		title := ""
		videoURL := ""
		meta := ""
		if v, ok := m["title"].(string); ok {
			title = v
		}
		if v, ok := m["url"].(string); ok {
			videoURL = v
		}
		if v, ok := m["meta"].(string); ok {
			meta = v
		}

		line := fmt.Sprintf("%d. %s", i+1, title)
		if meta != "" {
			line += " - " + meta
		}
		line += "\n   " + videoURL
		b.WriteString(line)
		if i < len(items)-1 {
			b.WriteString("\n")
		}
	}

	return R(b.String())
}

// ── BrowserScreenshotTool ─────────────────────────────────────────────────────

type BrowserScreenshotTool struct{}

func (t *BrowserScreenshotTool) Name() string { return "browser_screenshot" }

func (t *BrowserScreenshotTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Open a URL in a headless browser and capture a full-page screenshot.",
		Params: map[string]any{
			"url": prop("string", "URL to screenshot"),
		},
		Required: []string{"url"},
	}
}

func (t *BrowserScreenshotTool) Execute(args map[string]any) Result {
	rawURL := strings.TrimSpace(getStr(args, "url"))
	if !isSSRFSafeURL(rawURL) {
		return R("error: URL is blocked (SSRF or file scheme)")
	}

	_, tempPage, cleanup, err := newBrowserPage()
	if err != nil {
		return R("error: playwright not available — run `playwright install chromium`")
	}
	defer cleanup()

	if _, err := tempPage.Goto(rawURL, playwright.PageGotoOptions{Timeout: playwright.Float(20000)}); err != nil {
		return R("error: " + err.Error())
	}

	dir := screenshotDir()
	cleanupOldScreenshots(dir, 5*time.Minute)

	name := filepath.Join(dir, fmt.Sprintf("browser_%s.png", time.Now().Format("20060102_150405")))
	if _, err := tempPage.Screenshot(playwright.PageScreenshotOptions{
		Path:     playwright.String(name),
		FullPage: playwright.Bool(true),
	}); err != nil {
		return R("error: " + err.Error())
	}

	return R("saved: " + name)
}

// ── BrowserClickTool ──────────────────────────────────────────────────────────

type BrowserClickTool struct{}

func (t *BrowserClickTool) Name() string { return "browser_click" }

func (t *BrowserClickTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Click an element matching a CSS selector in the headless browser session.",
		Params: map[string]any{
			"selector": prop("string", "CSS selector or text of element to click"),
			"url":      prop("string", "Optional URL to navigate to first"),
			"timeout":  prop("integer", "Timeout in seconds (1-60, default 15)"),
		},
		Required: []string{"selector"},
	}
}

func (t *BrowserClickTool) Execute(args map[string]any) Result {
	sel := strings.TrimSpace(getStr(args, "selector"))
	if sel == "" {
		return R("error: selector is required")
	}

	rawURL := strings.TrimSpace(getStr(args, "url"))
	if rawURL != "" && !isSSRFSafeURL(rawURL) {
		return R("error: URL is blocked (SSRF or file scheme)")
	}

	timeoutSec := clampInt(getInt(args, "timeout", 15), 1, 60)
	timeoutMs := float64(timeoutSec * 1000)

	p, err := getPage()
	if err != nil {
		return R("error: playwright not available — run `playwright install chromium`")
	}

	if rawURL != "" {
		if _, err := p.Goto(rawURL, playwright.PageGotoOptions{Timeout: &timeoutMs}); err != nil {
			return R("error: " + err.Error())
		}
	}

	if err := p.Click(pwSelector(sel), playwright.PageClickOptions{Timeout: &timeoutMs}); err != nil {
		return R("error: " + err.Error())
	}

	p.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
		State:   playwright.LoadStateNetworkidle,
		Timeout: playwright.Float(5000),
	})

	snap := pageSnapshot(p)
	return R("clicked: " + sel + "\n" + snap)
}

// ── BrowserFillTool ───────────────────────────────────────────────────────────

type BrowserFillTool struct{}

func (t *BrowserFillTool) Name() string { return "browser_fill" }

func (t *BrowserFillTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Fill an input matching a CSS selector in the headless browser session.",
		Params: map[string]any{
			"selector": prop("string", "CSS selector or text of the input"),
			"value":    prop("string", "Value to fill"),
			"url":      prop("string", "Optional URL to navigate to first"),
			"submit":   prop("boolean", "Press Enter after filling"),
			"timeout":  prop("integer", "Timeout in seconds (1-60, default 15)"),
		},
		Required: []string{"selector", "value"},
	}
}

func (t *BrowserFillTool) Execute(args map[string]any) Result {
	sel := strings.TrimSpace(getStr(args, "selector"))
	if sel == "" {
		return R("error: selector is required")
	}
	value := getStr(args, "value")

	rawURL := strings.TrimSpace(getStr(args, "url"))
	if rawURL != "" && !isSSRFSafeURL(rawURL) {
		return R("error: URL is blocked (SSRF or file scheme)")
	}

	submit := false
	if v, ok := args["submit"].(bool); ok {
		submit = v
	}

	timeoutSec := clampInt(getInt(args, "timeout", 15), 1, 60)
	timeoutMs := float64(timeoutSec * 1000)

	p, err := getPage()
	if err != nil {
		return R("error: playwright not available — run `playwright install chromium`")
	}

	if rawURL != "" {
		if _, err := p.Goto(rawURL, playwright.PageGotoOptions{Timeout: &timeoutMs}); err != nil {
			return R("error: " + err.Error())
		}
	}

	if err := p.Fill(pwSelector(sel), value, playwright.PageFillOptions{Timeout: &timeoutMs}); err != nil {
		return R("error: " + err.Error())
	}

	if submit {
		if err := p.Press(pwSelector(sel), "Enter"); err != nil {
			return R("error: " + err.Error())
		}
		p.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
			State:   playwright.LoadStateNetworkidle,
			Timeout: playwright.Float(5000),
		})
	}

	snap := pageSnapshot(p)
	return R("filled: " + sel + "\n" + snap)
}

// playwrightCheck returns true if playwright is installed.
func playwrightCheck() bool {
	_, err := os.Stat(filepath.Join(os.Getenv("LOCALAPPDATA"), "ms-playwright"))
	return err == nil
}
